From 8e040390a9c2a113a1f40f69c57ee547c7aac918 Mon Sep 17 00:00:00 2001
From: Adam Duskett <Aduskett@rivian.com>
Date: Sun, 21 Feb 2021 16:13:34 -0800
Subject: [PATCH] fakeroot: new argument

allow users to specify a path to a fakeroot script. This allows users in a
cross-compiled environment to create nodes without needing access to the root
user.

Signed-off-by: Adam Duskett <Aduskett@rivian.com>
---
 dracut.sh | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/dracut.sh b/dracut.sh
index 885c921..3882e8f 100755
--- a/dracut.sh
+++ b/dracut.sh
@@ -103,6 +103,7 @@ Creates initial ramdisk images for preloading modules
                          initramfs.
   -k, --kmoddir [DIR]   Specify the directory, where to look for kernel
                          modules
+  --fakeroot [PATH]    Specify a path to a fakeroot binary. Useful for creating nodes in a cross-compiled environment.
   --fwdir [DIR]         Specify additional directories, where to look for
                          firmwares, separated by :
   --kernel-only         Only install kernel drivers and firmware files
@@ -353,6 +354,7 @@ rearrange_params()
         --long modules: \
         --long omit: \
         --long drivers: \
+        --long fakeroot: \
         --long filesystems: \
         --long install: \
         --long install-optional: \
@@ -528,6 +530,7 @@ while :; do
         -a|--add)      add_dracutmodules_l+=("$2");           PARMS_TO_STORE+=" '$2'"; shift;;
         --force-add)   force_add_dracutmodules_l+=("$2");     PARMS_TO_STORE+=" '$2'"; shift;;
         --add-drivers) add_drivers_l+=("$2");                 PARMS_TO_STORE+=" '$2'"; shift;;
+        --fakeroot)     fakeroot_path="$2"; shift;;
         --force-drivers) force_drivers_l+=("$2");             PARMS_TO_STORE+=" '$2'"; shift;;
         --omit-drivers) omit_drivers_l+=("$2");               PARMS_TO_STORE+=" '$2'"; shift;;
         -m|--modules)  dracutmodules_l+=("$2");               PARMS_TO_STORE+=" '$2'"; shift;;
@@ -1720,7 +1723,13 @@ if [[ $kernel_only != yes ]]; then
         # shellcheck disable=SC2174
         mkdir -m 0755 -p "${initdir}/lib/dracut/hooks/$_d"
     done
-    if [[ "$EUID" = "0" ]]; then
+    if [[ -n "${fakeroot_path}" ]]; then
+        [[ -c ${initdir}/dev/null ]] || FAKEROOTDONTTRYCHOWN=1 ${fakeroot_path} -- mknod "${initdir}"/dev/null c 1 3
+        [[ -c ${initdir}/dev/kmsg ]] || FAKEROOTDONTTRYCHOWN=1 ${fakeroot_path} -- mknod "${initdir}"/dev/kmsg c 1 11
+        [[ -c ${initdir}/dev/console ]] || FAKEROOTDONTTRYCHOWN=1 ${fakeroot_path} -- mknod "${initdir}"/dev/console c 5 1
+        [[ -c ${initdir}/dev/random ]] || FAKEROOTDONTTRYCHOWN=1 ${fakeroot_path} -- mknod "${initdir}"/dev/random c 1 8
+        [[ -c ${initdir}/dev/urandom ]] || FAKEROOTDONTTRYCHOWN=1 ${fakeroot_path} -- mknod "${initdir}"/dev/urandom c 1 9
+    elif [[ "$EUID" = "0" ]]; then
         [[ -c ${initdir}/dev/null ]] || mknod "${initdir}"/dev/null c 1 3
         [[ -c ${initdir}/dev/kmsg ]] || mknod "${initdir}"/dev/kmsg c 1 11
         [[ -c ${initdir}/dev/console ]] || mknod "${initdir}"/dev/console c 5 1
-- 
2.25.1

